{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree\n",
    "import pandas as pd\n",
    "import glob\n",
    "from Bio import SeqIO\n",
    "import random\n",
    "from random import choices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Function for getting closest relatives of one leaf \n",
    "### (It means getting all descendants of a parental node of a leaf)\n",
    "### Not used in the end\n",
    "def getRelatedLeafs(node):\n",
    "    nodeup = node.up\n",
    "    n = nodeup.get_descendants()\n",
    "    N = []\n",
    "    for nod in n:\n",
    "        if nod.is_leaf():\n",
    "            N.append(nod.name)\n",
    "    return N, nodeup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Reading population info and IR coordinates on the genome (.bed)\n",
    "samples_fn = \"./data/Species_distinction_PCA.txt\"\n",
    "df = pd.read_csv(samples_fn,sep='\\t',header=0)\n",
    "D = {a: [b,c] for a,b,c in zip(list(df[\"SpName\"]),list(df[\"GroupSamtools\"]),list(df[\"SamName\"]))}\n",
    "bed = pd.read_csv(\"./data/strain_merged_all_location.bed\", sep = \"\\t\", header = None, \n",
    "                  names = [\"chrom\",\"start\",\"stop\",\"strain\",\"lineage\",\"location\",\"region\"],\n",
    "                 dtype={\"start\": \"Int64\", \"stop\": \"Int64\"})\n",
    "ulmi = df.loc[df[\"GroupSamtools\"] == \"U\",\"SamName\"].values\n",
    "ulm = list(ulmi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Reading trees generated for each IR\n",
    "trees = glob.glob(\"./raxml_trees/RAxML_bipartitions_75coll.si_*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('TR88',\n",
       "  'OphioH327chr_6',\n",
       "  '1600000',\n",
       "  '2800000',\n",
       "  '785401_U,H1018_U,H1021_U,H1022_U,H1096_U,H2062_U,H337_U,H359_U,H379_U,H649_U,H809_U,H864_U,H868_U,H871_U,H872_U,H873_U,H909_U,H924_U,I131_U,I185_U,PG40_U'),\n",
       " ('R108',\n",
       "  'OphioH327chr_6',\n",
       "  '400000',\n",
       "  '600000',\n",
       "  '785401_U,H1018_U,H1021_U,H1022_U,H1096_U,H2062_U,H337_U,H359_U,H379_U,H649_U,H809_U,H864_U,H868_U,H871_U,H872_U,H873_U,H909_U,H924_U,I131_U,I185_U,PG40_U')]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### For each ONU strain with IR select all OU strains\n",
    "F = []\n",
    "for tree in trees:\n",
    "    strain = tree.split(\"_\")[6]\n",
    "    region = \"_\".join(tree.split(\"_\")[4:6])\n",
    "    #print(strain, region)\n",
    "    #t = Tree(tree,format=2)\n",
    "    #t.get_leaf_names()\n",
    "    #strainName = D[strain][1]\n",
    "    #strainNode = t&strainName\n",
    "    #relatives, uppernode = getRelatedLeafs(strainNode)\n",
    "    #leaf_nodes = [ele for ele in relatives if ele != strainName]\n",
    "    chrom = bed[(bed[\"start\"] == int(region.split(\"_\")[0])) & (bed[\"stop\"] == int(region.split(\"_\")[1])) & (bed[\"strain\"] == strain)]\n",
    "    f = strain, list(chrom['chrom'])[0], region.split(\"_\")[0], region.split(\"_\")[1], ','.join(ulm)\n",
    "    F.append(f)\n",
    "F[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Function for filtering sequences with too much missing data\n",
    "def filterBadSeq(BigList):\n",
    "    NewList = []\n",
    "    for sek in BigList:\n",
    "        length = len(sek)\n",
    "        nmissi = sek.count('N')\n",
    "        propmi = nmissi/float(length)\n",
    "        if propmi > 0.75:\n",
    "            continue\n",
    "        else:\n",
    "            NewList.append(sek)\n",
    "    return NewList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Reading fasta alignments of IRs; for each introgressed ONU removing poor quality sequences \n",
    "### and selecting only the this ONU strain, one OU sequences, and O. himal-ulmi sequences (outgroup)\n",
    "T = {}\n",
    "for ele in F:\n",
    "    ulmi = ele[-1].split(\",\")\n",
    "    fnames = glob.glob(\"./fasta/si_\"+ele[2]+\"_\"+ele[3]+\"_\"+ele[0]+\"_*.fasta\")\n",
    "    N,O,U = [],[],[]\n",
    "    for record in SeqIO.parse(fnames[0],\"fasta\"):\n",
    "        if record.id == D[ele[0]][1]:\n",
    "            N.append(str(record.seq).upper())\n",
    "        elif record.id in [\"HP30_H\",\"HP31_H\",\"HP32_H\"]:\n",
    "            O.append(str(record.seq).upper())\n",
    "        elif record.id in ulmi:\n",
    "            U.append(str(record.seq).upper())\n",
    "    \n",
    "    dN = filterBadSeq(N)\n",
    "    booln = 1\n",
    "    if dN == []:\n",
    "        print(ele[0]+\"_\"+ele[1]+\"_\"+ele[2]+\"_\"+ele[3]+\" novo-ulmi too bad\")\n",
    "        booln = 0\n",
    "    dO = filterBadSeq(O)\n",
    "    if dO == []:\n",
    "        print(ele[0]+\"_\"+ele[1]+\"_\"+ele[2]+\"_\"+ele[3]+\" no outgroup\")\n",
    "        booln = 0\n",
    "    dU = filterBadSeq(U)\n",
    "    if dU == []:\n",
    "        print(ele[0]+\"_\"+ele[1]+\"_\"+ele[2]+\"_\"+ele[3]+\" no ulmi strain\")\n",
    "        booln = 0\n",
    "    \n",
    "    if booln == 1:\n",
    "        outgroup_tuple = list(zip(*dO))\n",
    "        focal_nt = list(dN[0])\n",
    "        ulmi_tuple = list(zip(*dU))\n",
    "        T[ele[0]+\"_\"+ele[1]+\"_\"+ele[2]+\"_\"+ele[3]] = outgroup_tuple, focal_nt, ulmi_tuple\n",
    "        \n",
    "#list(T.keys())[:4]\n",
    "#T['TR88_OphioH327chr_6_1600000_2800000']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TR88_OphioH327chr_6_1600000_2800000\n",
      "R108_OphioH327chr_6_400000_600000\n",
      "R108_OphioH327chr_3_0_200000\n",
      "V19_OphioH327chr_7_0_200000\n",
      "Yu16_OphioH327chr_1_0_200000\n",
      "H371_OphioH327chr_1_0_400000\n",
      "H986_OphioH327chr_1_0_400000\n",
      "P122_OphioH327chr_1_0_400000\n",
      "V19_OphioH327chr_1_0_400000\n",
      "TR88_OphioH327chr_1_1600000_2200000\n",
      "H413a_OphioH327chr_8_2000000_2600000\n",
      "R22_OphioH327chr_1_200000_400000\n",
      "P122_OphioH327chr_4_2200000_2400000\n",
      "V19_OphioH327chr_7_2200000_2400000\n",
      "H323_OphioH327chr_5_2400000_2600000\n",
      "TR88_OphioH327chr_3_3600000_3800000\n",
      "V19_OphioH327chr_3_3600000_3800000\n",
      "H276_OphioH327chr_6_400000_600000\n",
      "P114_OphioH327chr_5_400000_600000\n",
      "P122_OphioH327chr_6_400000_600000\n",
      "R22_OphioH327chr_6_400000_600000\n",
      "H2098_OphioH327chr_2_4800000_5800000\n",
      "H2091_OphioH327chr_2_5200000_5600000\n",
      "Yu16_OphioH327chr_2_5200000_5600000\n",
      "DDS298_OphioH327chr_2_5400000_5600000\n",
      "DDS93_OphioH327chr_2_5400000_5600000\n",
      "US137_OphioH327chr_2_5400000_5600000\n",
      "H413a_OphioH327chr_2_5400000_6800000\n",
      "H986_OphioH327chr_2_5800000_6800000\n",
      "R108_OphioH327chr_8_600000_1000000\n",
      "TR88_OphioH327chr_8_600000_1200000\n",
      "H986_OphioH327chr_8_600000_800000\n",
      "V19_OphioH327chr_8_600000_800000\n",
      "Yu141_OphioH327chr_8_600000_800000\n",
      "H323_OphioH327chr_2_6200000_6600000\n"
     ]
    }
   ],
   "source": [
    "### For each IR, after removing 20 kb on both sides of the IR, select random bases for OU and outgroup,\n",
    "### and encode each base as a 3-number code\n",
    "TN = {}\n",
    "AS = {}\n",
    "edges_correction=True\n",
    "for ele in list(T.keys()):\n",
    "    print(ele)\n",
    "    \n",
    "    if edges_correction == True:\n",
    "        outgroup = T[ele][0][20000:-20000]\n",
    "        focal = T[ele][1][20000:-20000]\n",
    "        ulmi = T[ele][2][20000:-20000]\n",
    "        N010,N001, N, N011 = [],[],[],[]\n",
    "        t = []\n",
    "    else:\n",
    "        outgroup = T[ele][0]\n",
    "        focal = T[ele][1]\n",
    "        ulmi = T[ele][2]\n",
    "        N010,N001, N, N011 = [],[],[],[]\n",
    "        t = []\n",
    "    pats = []\n",
    "    for nt in range(len(focal)):\n",
    "        out_flt = [i for i in list(outgroup[nt]) if i != 'N']\n",
    "        focal_flt = [i for i in [focal[nt]] if i != 'N']\n",
    "        ulmi_flt = [i for i in list(ulmi[nt]) if i != 'N']\n",
    "        if (out_flt!=[]) & (focal_flt!=[]) & (ulmi_flt!=[]):\n",
    "            N.append(1)\n",
    "            out = random.sample(out_flt,1)[0]\n",
    "            foc = focal_flt[0]\n",
    "            ulm = random.sample(ulmi_flt,1)[0]\n",
    "            if (out == ulm) & (out!=foc):\n",
    "                N010.append(1)\n",
    "                pats.append('010')\n",
    "            elif (out != ulm) & (out==foc):\n",
    "                N001.append(1)\n",
    "                pats.append('001')\n",
    "            elif (out != ulm) & (out != foc):\n",
    "                N011.append(1)\n",
    "                pats.append('011')\n",
    "            else:\n",
    "                pats.append('000')\n",
    "        t.append([out,foc,ulm])\n",
    "\n",
    "    TN[ele] = sum(N),sum(N010),sum(N001), sum(N011)\n",
    "    AS[ele] = pats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Saving average branch length estimates\n",
    "bD = {}\n",
    "for ele in TN.keys():\n",
    "    bD[ele] = \"_\".join(ele.split(\"_\")[1:3]),int(ele.split(\"_\")[3]),int(ele.split(\"_\")[4]),ele.split(\"_\")[0],TN[ele][0],TN[ele][1],TN[ele][2],TN[ele][3]\n",
    "\n",
    "x1,x2,x3,x4,x5,x6,x7,x8 = list(zip(*list(bD.values())))\n",
    "dbd = pd.DataFrame({\"chrom\":x1,\"start\":x2,\"stop\":x3,\"strain\":x4,\"allSites\":x5,\"focSites\":x6,\"ulmSites\":x7,\"fixedHIM\":x8})\n",
    "dx = pd.merge(dbd,bed,on = [\"chrom\",\"start\",\"stop\",\"strain\"], how = \"left\")\n",
    "dx[\"Length\"]=dx[\"stop\"]-dx[\"start\"]\n",
    "dx[\"ttout\"]=((dx[\"focSites\"] + dx[\"ulmSites\"])/2)/dx[\"fixedHIM\"]\n",
    "#dx.head()\n",
    "dx.to_csv(\"./data/selectStrainsForbranchEstimationAllOU.out\",sep = \"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TR88_OphioH327chr_6_1600000_2800000\n",
      "R108_OphioH327chr_6_400000_600000\n",
      "R108_OphioH327chr_3_0_200000\n",
      "V19_OphioH327chr_7_0_200000\n",
      "Yu16_OphioH327chr_1_0_200000\n",
      "H371_OphioH327chr_1_0_400000\n",
      "H986_OphioH327chr_1_0_400000\n",
      "P122_OphioH327chr_1_0_400000\n",
      "V19_OphioH327chr_1_0_400000\n",
      "TR88_OphioH327chr_1_1600000_2200000\n",
      "H413a_OphioH327chr_8_2000000_2600000\n",
      "R22_OphioH327chr_1_200000_400000\n",
      "P122_OphioH327chr_4_2200000_2400000\n",
      "V19_OphioH327chr_7_2200000_2400000\n",
      "H323_OphioH327chr_5_2400000_2600000\n",
      "TR88_OphioH327chr_3_3600000_3800000\n",
      "V19_OphioH327chr_3_3600000_3800000\n",
      "H276_OphioH327chr_6_400000_600000\n",
      "P114_OphioH327chr_5_400000_600000\n",
      "P122_OphioH327chr_6_400000_600000\n",
      "R22_OphioH327chr_6_400000_600000\n",
      "H2098_OphioH327chr_2_4800000_5800000\n",
      "H2091_OphioH327chr_2_5200000_5600000\n",
      "Yu16_OphioH327chr_2_5200000_5600000\n",
      "DDS298_OphioH327chr_2_5400000_5600000\n",
      "DDS93_OphioH327chr_2_5400000_5600000\n",
      "US137_OphioH327chr_2_5400000_5600000\n",
      "H413a_OphioH327chr_2_5400000_6800000\n",
      "H986_OphioH327chr_2_5800000_6800000\n",
      "R108_OphioH327chr_8_600000_1000000\n",
      "TR88_OphioH327chr_8_600000_1200000\n",
      "H986_OphioH327chr_8_600000_800000\n",
      "V19_OphioH327chr_8_600000_800000\n",
      "Yu141_OphioH327chr_8_600000_800000\n",
      "H323_OphioH327chr_2_6200000_6600000\n"
     ]
    }
   ],
   "source": [
    "### Run 100 empirical bootstraps of base patters for each IR\n",
    "boot = 100\n",
    "BT = {}\n",
    "for ele in list(AS.keys()):\n",
    "    print(ele)\n",
    "    A = len(AS[ele])\n",
    "    L = A\n",
    "    #L = minlen\n",
    "    Boot = []\n",
    "    for bootstrap in range(boot):\n",
    "        bootsites = random.choices(AS[ele],k=int(L))\n",
    "        n010 = bootsites.count(\"010\")\n",
    "        n001 = bootsites.count(\"001\")\n",
    "        n011 = bootsites.count(\"011\")\n",
    "        n000 = bootsites.count(\"000\")\n",
    "        tot = n010 + n001 + n011 + n000\n",
    "        ttout = ((n010+n001)/2)/n011\n",
    "        Boot.append([bootstrap,n010,n001,n011,n000,tot,ttout])\n",
    "    BT[ele] = Boot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>length</th>\n",
       "      <th>boot</th>\n",
       "      <th>n010</th>\n",
       "      <th>n001</th>\n",
       "      <th>n011</th>\n",
       "      <th>n000</th>\n",
       "      <th>tot</th>\n",
       "      <th>ttout</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>0</td>\n",
       "      <td>1491</td>\n",
       "      <td>1694</td>\n",
       "      <td>47552</td>\n",
       "      <td>904819</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.033490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1</td>\n",
       "      <td>1558</td>\n",
       "      <td>1717</td>\n",
       "      <td>47400</td>\n",
       "      <td>904881</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.034546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>2</td>\n",
       "      <td>1483</td>\n",
       "      <td>1743</td>\n",
       "      <td>47569</td>\n",
       "      <td>904761</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.033909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>3</td>\n",
       "      <td>1509</td>\n",
       "      <td>1694</td>\n",
       "      <td>47390</td>\n",
       "      <td>904963</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.033794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>4</td>\n",
       "      <td>1506</td>\n",
       "      <td>1658</td>\n",
       "      <td>47687</td>\n",
       "      <td>904705</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.033175</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  strain           chrom    start     stop   length  boot  n010  n001   n011  \\\n",
       "0   TR88  OphioH327chr_6  1600000  2800000  1200000     0  1491  1694  47552   \n",
       "1   TR88  OphioH327chr_6  1600000  2800000  1200000     1  1558  1717  47400   \n",
       "2   TR88  OphioH327chr_6  1600000  2800000  1200000     2  1483  1743  47569   \n",
       "3   TR88  OphioH327chr_6  1600000  2800000  1200000     3  1509  1694  47390   \n",
       "4   TR88  OphioH327chr_6  1600000  2800000  1200000     4  1506  1658  47687   \n",
       "\n",
       "     n000     tot     ttout  \n",
       "0  904819  955556  0.033490  \n",
       "1  904881  955556  0.034546  \n",
       "2  904761  955556  0.033909  \n",
       "3  904963  955556  0.033794  \n",
       "4  904705  955556  0.033175  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Create data frame with bootstrap results\n",
    "R = []\n",
    "for ele in BT.keys():\n",
    "    strain = ele.split(\"_\")[0]\n",
    "    chrom = \"_\".join(ele.split(\"_\")[1:3])\n",
    "    start = ele.split(\"_\")[3]\n",
    "    stop = ele.split(\"_\")[4]\n",
    "    length = int(stop) - int(start)\n",
    "    for boot in BT[ele]:\n",
    "        r = [strain, chrom, int(start), int(stop), str(length)] + boot\n",
    "        R.append(r)\n",
    "Rzip = list(zip(*R))\n",
    "dr = pd.DataFrame({\"strain\":Rzip[0],\"chrom\":Rzip[1],\"start\":Rzip[2],\"stop\":Rzip[3],\"length\":Rzip[4],\"boot\":Rzip[5],\"n010\":Rzip[6],\"n001\":Rzip[7],\"n011\":Rzip[8],\"n000\":Rzip[9],\"tot\":Rzip[10],\"ttout\":Rzip[11]})\n",
    "dr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain_x</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>length</th>\n",
       "      <th>boot</th>\n",
       "      <th>n010</th>\n",
       "      <th>n001</th>\n",
       "      <th>n011</th>\n",
       "      <th>n000</th>\n",
       "      <th>tot</th>\n",
       "      <th>ttout</th>\n",
       "      <th>strain_y</th>\n",
       "      <th>lineage</th>\n",
       "      <th>location</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>0</td>\n",
       "      <td>1491</td>\n",
       "      <td>1694</td>\n",
       "      <td>47552</td>\n",
       "      <td>904819</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.033490</td>\n",
       "      <td>TR88</td>\n",
       "      <td>NOV</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1</td>\n",
       "      <td>1558</td>\n",
       "      <td>1717</td>\n",
       "      <td>47400</td>\n",
       "      <td>904881</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.034546</td>\n",
       "      <td>TR88</td>\n",
       "      <td>NOV</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>2</td>\n",
       "      <td>1483</td>\n",
       "      <td>1743</td>\n",
       "      <td>47569</td>\n",
       "      <td>904761</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.033909</td>\n",
       "      <td>TR88</td>\n",
       "      <td>NOV</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>3</td>\n",
       "      <td>1509</td>\n",
       "      <td>1694</td>\n",
       "      <td>47390</td>\n",
       "      <td>904963</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.033794</td>\n",
       "      <td>TR88</td>\n",
       "      <td>NOV</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR88</td>\n",
       "      <td>OphioH327chr_6</td>\n",
       "      <td>1600000</td>\n",
       "      <td>2800000</td>\n",
       "      <td>1200000</td>\n",
       "      <td>4</td>\n",
       "      <td>1506</td>\n",
       "      <td>1658</td>\n",
       "      <td>47687</td>\n",
       "      <td>904705</td>\n",
       "      <td>955556</td>\n",
       "      <td>0.033175</td>\n",
       "      <td>TR88</td>\n",
       "      <td>NOV</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  strain_x           chrom    start     stop   length  boot  n010  n001  \\\n",
       "0     TR88  OphioH327chr_6  1600000  2800000  1200000     0  1491  1694   \n",
       "1     TR88  OphioH327chr_6  1600000  2800000  1200000     1  1558  1717   \n",
       "2     TR88  OphioH327chr_6  1600000  2800000  1200000     2  1483  1743   \n",
       "3     TR88  OphioH327chr_6  1600000  2800000  1200000     3  1509  1694   \n",
       "4     TR88  OphioH327chr_6  1600000  2800000  1200000     4  1506  1658   \n",
       "\n",
       "    n011    n000     tot     ttout strain_y lineage  location  region  \n",
       "0  47552  904819  955556  0.033490     TR88     NOV        12      10  \n",
       "1  47400  904881  955556  0.034546     TR88     NOV        12      10  \n",
       "2  47569  904761  955556  0.033909     TR88     NOV        12      10  \n",
       "3  47390  904963  955556  0.033794     TR88     NOV        12      10  \n",
       "4  47687  904705  955556  0.033175     TR88     NOV        12      10  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Append info about location number\n",
    "dm = pd.merge(dr, bed, on = [\"chrom\",\"start\",\"stop\"])\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save results to table\n",
    "dm.to_csv(\"./data/selectStrainsForbranchEstimationBootstrapAllOU.out\",sep = \"\\t\", header=True, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
